File: frontend/src/types/learning.ts
export type ExerciseType = 'multiple_choice' | 'fill_in_blank' | 'free_text'
export type Difficulty = 'easy' | 'medium' | 'hard'

export interface Exercise {
  id: number
  entity_id: number
  question: string
  exercise_type: ExerciseType
  difficulty: Difficulty
  bloom_level: number
  options_json: { options: string[] } | null
  is_answered: boolean
  is_correct: boolean | null
  correct_answer: string | null
  explanation: string | null
  user_answer: string | null
  score: number | null
  created_at: string
  answered_at: string | null
}

export interface AnswerRequest {
  user_answer: string
  rating: number
}

export interface LearningStats {
  total_entities: number
  mastered_entities: number
  average_mastery: number
  total_exercises: number
  answered_exercises: number
  correct_exercises: number
  accuracy: number
}

export interface Streak {
  current_streak: number
  longest_streak: number
  last_activity_date: string | null
  total_active_days: number
  multiplier: number
}

export interface LearningPriority {
  id: number
  entity_id: number
  composite_score: number
  deadline_urgency: number
  topic_relevance: number
  mastery_gap: number
  decay_amount: number
  bloom_gap: number
  is_blocked: boolean
  block_reason: string | null
  calculated_at: string
}

File: backend/app/schemas/learning.py
from datetime import date, datetime

from pydantic import BaseModel


class ExerciseOut(BaseModel):
    model_config = {"from_attributes": True}

    id: int
    entity_id: int
    question: str
    exercise_type: str
    difficulty: str
    bloom_level: int
    options_json: dict | None = None
    is_answered: bool
    is_correct: bool | None = None
    correct_answer: str | None = None
    explanation: str | None = None
    user_answer: str | None = None
    score: float | None = None
    created_at: datetime
    answered_at: datetime | None = None


class AnswerRequest(BaseModel):
    user_answer: str
    rating: int  # 1=Again, 2=Hard, 3=Good, 4=Easy


class ExerciseRequest(BaseModel):
    entity_id: int
    difficulty: str = "medium"
    bloom_level: int | None = None


class SessionRequest(BaseModel):
    count: int = 10


class StatsOut(BaseModel):
    total_entities: int
    mastered_entities: int
    average_mastery: float
    total_exercises: int
    answered_exercises: int
    correct_exercises: int
    accuracy: float


class StreakOut(BaseModel):
    model_config = {"from_attributes": True}

    current_streak: int
    longest_streak: int
    last_activity_date: date | None = None
    total_active_days: int
    multiplier: float = 1.0


class PriorityOut(BaseModel):
    model_config = {"from_attributes": True}

    id: int
    entity_id: int
    composite_score: float
    deadline_urgency: float
    topic_relevance: float
    mastery_gap: float
    decay_amount: float
    bloom_gap: float
    is_blocked: bool
    block_reason: str | None = None
    calculated_at: datetime
